### PR TITLE
add audiobookshelf install sh

### PR DESCRIPTION
**Pull request by lupu60**
Add installation for audiobookshelf

## Description
Add installation for audiobookshelf

## Change Categories
- New feature 

## Checklist

- [x] Branch was made off the `develop` branch and the PR is targetting the `develop` branch
- [ ] Docs have been made OR are not necessary
    - PR link: 
- [ ] Changes to panel have been made OR are not necessary
    - PR link: 
- [x] Code is formatted [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Code conforms to project structure [(See more)](https://swizzin.ltd/dev/structure)
- [ ] Shellcheck isn't screaming [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [ ] Prints to terminal are handled [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#printing-into-the-terminal)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] Testing was done
   - [ ] Tests created or no new tests necessary
   - [ ] Tests executed

### Architectures

|   			| `amd64` 	| `armhf` 	| `arm64` 	| Unspecified 	|
|--------		|-------- 	|-------- 	|-------- 	|----------		|
| Jammy 		|			|			|			|				|
| Focal 		|			|			|			|				|
| Bionic		|			|			|			|				|
| Bullseye		|			|			|			|				|
| Buster		|			|			|			|				|
| Stretch		|			|			|			|				|
| Raspbian  	|	⚫️		|			|	⚫️		|	⚫️			|

### CODE REVIEW

**Comment by liaralabs** [comment]
Thanks for the contribution lupu60. I have a few items of feedback:

- At the moment we are requiring contributors of new applications to be willing to maintain these applications long term. The core maintainers are already stretched thin with maintaining the existing packages. Are you willing to commit to this?
- The package is missing nginx proxy configs
- Dashboard integration: You should submit a PR to the swizzin_dashboard repository to make this changes there rather than echoing these changes into the profiles. The method you have chosen is not guaranteed to not be clobbered during box updates

**Comment by lupu60** [comment]
> 

Hello @liaralabs.
* I can handle audiobookshelf and I'm sure others will join after will be available.
* I need to investigate the nginx config, looks like there is an issue:
   https://github.com/advplyr/audiobookshelf/issues/385
* I will adapt the pr and open a new one in the swizzin_dashboard

File: scripts/install/audiobookshelf.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    systemd = "audiobookshelf"\n
+    check_theD = True'
+
+# Write content to the file
+echo -e "$content" > /opt/swizzin/core/custom/profiles.py
+
+echo "Content has been written to /opt/swizzin/core/custom/profiles.py"
+touch /install/.audiobookshelf.lock
+systemctl restart panel
+}
```
Comment (flying-sausages) [inline scripts/install/audiobookshelf.sh:33]: Could you remove this and make a PR on the panel repo please? Let me know if you need help in the discord

File: scripts/install/audiobookshelf.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+# Audiobookshelf installer
+
+_install_audiobookshelf() {
+    echo_progress_start "Installing audiobookshelf apt sources"
```
Comment (flying-sausages) [inline scripts/install/audiobookshelf.sh:5]: `echo_progress_start` kind of expects an `echo_progress_done`, think of the two as the slices of bread in a waiting sandwich

File: scripts/install/audiobookshelf.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    apt install gnupg curl
```
Comment (flying-sausages) [inline scripts/install/audiobookshelf.sh:6]: We have `apt_install` defined which helps handle these without the interactivity

File: scripts/install/audiobookshelf.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    curl -s https://advplyr.github.io/audiobookshelf-ppa/KEY.gpg |  apt-key add -
+    curl -s -o /etc/apt/sources.list.d/audiobookshelf.list https://advplyr.github.io/audiobookshelf-ppa/audiobookshelf.list
+    apt update
+    apt install audiobookshelf
```
Comment (flying-sausages) [inline scripts/install/audiobookshelf.sh:10]: Same as above, would be nice to keep the apt installs to one call with one big list in my opinion

File: scripts/install/audiobookshelf.sh (hunk 3 above)
Comment (flying-sausages) [reply scripts/install/audiobookshelf.sh:6]: You can use it like this, it is imported into the environment anytime you run `box`, you can check `sources/globals.sh` to see what else there is
```suggestion
    apt_install gnupg curl
```

File: scripts/remove/audiobookshelf.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    rm -f /etc/apt/sources.list.d/audiobookshelf.list
+    apt update
+    apt autoremove
+    rm -f /opt/swizzin/static/img/apps/audiobookshelf.png
+}
+
+__remove_from_pannel() {
+    content= 'from core.profiles import *\n\n'
+    echo -e "$content" > /opt/swizzin/core/custom/profiles.py
+}
```
Comment (flying-sausages) [inline scripts/remove/audiobookshelf.sh:18]: Again, not really needed if it's in the panel repo

File: scripts/remove/audiobookshelf.sh [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# Audiobookshelf
+
+__remove_audiobookshelf() {
+    echo_progress_start "Removing audiobookshelf apt sources"
+    systemctl stop audiobookshelf.service
+    sysctl disable audiobookshelf.service
+    apt remove audiobookshelf
+    rm -f /etc/apt/sources.list.d/audiobookshelf.list
+    apt update
+    apt autoremove
```
Comment (flying-sausages) [inline scripts/remove/audiobookshelf.sh:11]: We've got apt functions for these two, but I don't think you need to do the update here, just takes time. The apt tools will auto-update themselves as they're being called if necessary

File: scripts/remove/audiobookshelf.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+# Audiobookshelf
+
+__remove_audiobookshelf() {
+    echo_progress_start "Removing audiobookshelf apt sources"
+    systemctl stop audiobookshelf.service
+    sysctl disable audiobookshelf.service
```
Comment (flying-sausages) [inline scripts/remove/audiobookshelf.sh:7]: `disable` also takes `--now` parameter so you can get these done in one line. Would also be nice to pass the `-q` flag so it's quiet

**Review by flying-sausages — CHANGES_REQUESTED** [review]
Mostly what Liara said, let me know if there's anything I can help you with!

**Comment by ligix** [comment]
> At the moment we are requiring contributors of new applications to be willing to maintain these applications long term. The core maintainers are already stretched thin with maintaining the existing packages. Are you willing to commit to this?

I would but sadly ABS uses the header `Authorization` which prevents the browser from sending the basic auth credentials. This, for obvious reasons, breaks swizzin and the only workaround appears to disable authorization altogether.

For anyone wandering over from google this works:

/etc/nginx/apps/audiobookself.conf
```
location /abs {
    return 301 /audiobookshelf/;
}

location /audiobookshelf/ {
    proxy_pass              http://127.0.0.1:13378;
    proxy_http_version      1.1;
    proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_set_header Host               $http_host;
    proxy_set_header Upgrade            $http_upgrade;
    proxy_set_header Connection         "upgrade";
    proxy_set_header        X-Forwarded-Host        $http_host;
    proxy_redirect                      http:// https://;
    client_max_body_size                10240M;
}
```

**Pull request closed without merging**